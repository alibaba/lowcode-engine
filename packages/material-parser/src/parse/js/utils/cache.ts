export interface ICache {
  [name: string]: any;
}

const cache: ICache = {};

export function set(scope: string, name: string, value: any) {
  cache[scope] = cache[scope] || {};
  cache[scope][name] = value;
}

export function get(scope: string, name: string) {
  return (cache[scope] || {})[name];
}

export function has(scope: string, name: string) {
  return cache[scope] && Object.prototype.hasOwnProperty.call(cache[scope], name);
}
