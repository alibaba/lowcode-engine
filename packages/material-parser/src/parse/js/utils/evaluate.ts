const { namedTypes: t } = require('ast-types');

const { resolveToValue } = require('react-docgen').utils;

function isInfinity(path) {
  return t.Identifier.check(path.node) && path.node.name === 'Infinity';
}

function wrapValue(value, confident = true) {
  return {
    value,
    confident,
  };
}

export default function evaluate(path: any) {
  if (t.UnaryExpression.check(path.node)) {
    if (path.node.operator === 'void') {
      return wrapValue(undefined);
    }

    const argument = path.get('argument');

    const result = evaluate(argument);
    if (!result.confident) {
      return wrapValue(undefined, false);
    }

    const arg = result.value;
    if (arg === undefined) {
      return wrapValue(undefined);
    }

    switch (path.node.operator) {
      case '!':
        return wrapValue(!arg);
      case '+':
        return wrapValue(+arg);
      case '-':
        return wrapValue(-arg);
      case '~':
        return wrapValue(~arg);
      case 'typeof':
        return wrapValue(typeof arg);
    }
  }

  if (t.Identifier.check(path.node)) {
    const valuePath = resolveToValue(path);
    if (isInfinity(valuePath)) {
      return wrapValue(undefined);
    }
    return evaluate(valuePath);
  }

  if (t.Literal.check(path.node)) {
    return wrapValue(path.node.value);
  }

  if (t.ObjectExpression.check(path.node)) {
    const returnValue = {};
    path.get('properties').each((propertyPath) => {
      const { confident, value } = evaluate(propertyPath.get('value'));
      if (!confident) {
        return;
      }
      const keyPath = propertyPath.get('key');
      let key;
      if (keyPath.node.computed) {
        const result = evaluate(keyPath);
        if (!result.confident) {
          return;
        }
        key = result.value;
      } else {
        key = keyPath.node.name;
      }
      returnValue[key] = value;
    });

    return wrapValue(returnValue);
  }

  if (t.ArrayExpression.check(path.node)) {
    const value = [];
    let isValid = true;
    path.get('elements').each((x) => {
      if (!isValid) return;
      const result = evaluate(x);
      if (!result.confident) {
        isValid = false;
      } else {
        value.push(result.value);
      }
    });

    if (isValid) {
      return wrapValue(value);
    }
  }

  return wrapValue(undefined, false);
}
