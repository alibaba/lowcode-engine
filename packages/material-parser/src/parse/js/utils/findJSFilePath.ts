import fs from 'fs';

const suffixes = ['js', 'jsx'];
export default function findJSFilePath(fileBasePath: string): string {
  let filePath;
  for (const suffix of suffixes) {
    const fp = `${fileBasePath}.${suffix}`;
    if (fs.existsSync(fp)) {
      filePath = fp;
      break;
    }
  }

  return filePath;
}