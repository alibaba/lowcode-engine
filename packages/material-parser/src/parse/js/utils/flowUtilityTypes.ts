/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import { namedTypes as t } from 'ast-types';

const supportedUtilityTypes = new Set(['$Exact', '$ReadOnly']);

/**
 * See `supportedUtilityTypes` for which types are supported and
 * https://flow.org/en/docs/types/utilities/ for which types are available.
 */
function isSupportedUtilityType(path: any) {
  if (t.GenericTypeAnnotation.check(path.node)) {
    const idPath = path.get('id');
    return !!idPath && supportedUtilityTypes.has(idPath.node.name);
  }
  return false;
}

export { isSupportedUtilityType };

function isReactUtilityType(path: any) {
  if (t.TSTypeReference.check(path.node)) {
    const objName = path.get('typeName', 'left').node.name;
    if (objName === 'React') {
      return true;
    }
  }

  return false;
}

/**
 * Unwraps well known utility types. For example:
 *
 *   $ReadOnly<T> => T
 */
function unwrapUtilityType(path: any) {
  while (isSupportedUtilityType(path) || isReactUtilityType(path)) {
    path = path.get('typeParameters', 'params', 0);
  }

  return path;
}

export { unwrapUtilityType };
