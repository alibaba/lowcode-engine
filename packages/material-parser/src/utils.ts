import { pathExists, readFileSync, writeFile } from 'fs-extra';
import { isPlainObject } from 'lodash';
import safeEval from 'safe-eval';
import * as path from 'path';
import spawn from 'cross-spawn-promise';

export async function isNPMInstalled(args: {
  workDir: string;
  moduleDir: string;
  npmClient?: string;
}) {
  return pathExists(path.join(args.workDir, 'node_modules'));
}

export async function install(args: { workDir: string; moduleDir: string; npmClient?: string }) {
  if (await isNPMInstalled(args)) return;
  const { workDir, npmClient = 'tnpm' } = args;
  try {
    await spawn(npmClient, ['i'], { stdio: 'inherit', cwd: workDir } as any);
  } catch (e) {
    // TODO
  }
}

export async function installTypeScript(args: {
  workDir: string;
  moduleDir: string;
  npmClient?: string;
}) {
  if (await isNPMInstalled(args)) return;
  const { workDir, npmClient = 'tnpm' } = args;
  await spawn(npmClient, ['i', 'typescript'], { stdio: 'inherit', cwd: workDir } as any);
}

export async function installPeerDeps(args: {
  workDir: string;
  moduleDir: string;
  npmClient?: string;
}) {
  const { workDir, moduleDir, npmClient = 'tnpm' } = args;
  const modulePkgJsonPath = path.resolve(moduleDir, 'package.json');
  if (!(await pathExists(modulePkgJsonPath))) {
    return;
  }
  const pkgJsonPath = path.resolve(workDir, 'package.json');
  if (!(await pathExists(pkgJsonPath))) {
    return;
  }
  const modulePkgJson = await resolvePkgJson(modulePkgJsonPath);
  const pkgJson = await resolvePkgJson(pkgJsonPath);
  const { peerDependencies = {} } = modulePkgJson;
  pkgJson.dependencies = pkgJson.dependencies || {};
  pkgJson.dependencies = {
    ...pkgJson.dependencies,
    ...peerDependencies,
  };
  await writeFile(pkgJsonPath, JSON.stringify(pkgJson, null, 2));
  await spawn(npmClient, ['i'], { stdio: 'inherit', cwd: workDir } as any);
}

export async function installTypeModules(args: {
  workDir: string;
  moduleDir: string;
  npmClient?: string;
}) {
  const { workDir, moduleDir, npmClient = 'tnpm' } = args;
  const pkgJsonPath = path.resolve(moduleDir, 'package.json');
  if (!(await pathExists(pkgJsonPath))) {
    return;
  }
  await spawn(npmClient.replace('m', 'x'), ['typesync'], { stdio: 'inherit', cwd: workDir } as any);
}

export async function resolvePkgJson(pkgJsonPath: string): Promise<{ [k: string]: any }> {
  const content = await loadFile(pkgJsonPath);
  const json = JSON.parse(content);
  return json;
}

export function loadFile(filePath: string): string {
  const content: string | Buffer = readFileSync(filePath);
  if (typeof content === 'string') {
    return content;
  }
  return content.toString();
}

export function isPrimitive(val) {
  return !['object', 'function'].includes(typeof val) || val === null;
}

export function isEvaluable(value) {
  if (isPrimitive(value)) return true;
  if (Array.isArray(value)) {
    return value.every(isEvaluable);
  } else if (isPlainObject(value)) {
    return Object.keys(value).every(key => isEvaluable(value[key]));
  }
  return false;
}

export { safeEval };
