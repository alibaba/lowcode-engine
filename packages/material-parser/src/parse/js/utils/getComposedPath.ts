import { namedTypes as t, visit } from 'ast-types';
import fs from 'fs';
import path from 'path';
import getRoot from './getRoot';
import findJSFilePath from './findJSFilePath';

const buildParser = require('react-docgen/dist/babelParser').default;
const expressionTo = require('react-docgen/dist/utils/expressionTo');

const {
  resolveToValue,
  isExportsOrModuleAssignment,
} = require('react-docgen').utils;

export default function getComposedPropTypesPath(documentation, propName, p) {
  const composes: string[] = Array.from(documentation._composes);
  let _path = null;
  const root = getRoot(p).node;
  for (const compose of composes) {
    const composePath = findJSFilePath(path.resolve(path.dirname(root.__path), compose));
    if (!composePath) continue;

    const fileContent = fs.readFileSync(composePath, 'utf8');
    const parser = buildParser({ filename: composePath });
    const ast = parser.parse(fileContent);

    visit(ast, {
      visitAssignmentExpression(path: any) {
        // Ignore anything that is not `exports.X = ...;` or
        // `module.exports = ...;`
        if (!isExportsOrModuleAssignment(path)) {
          return false;
        }
        const arr = expressionTo.Array(path.get('left'));
        if (!(arr[0] === 'exports' && arr[1] === propName)) return false;

        // Resolve the value of the right hand side. It should resolve to a call
        // expression, something like React.createClass
        path = resolveToValue(path.get('right'));
        _path = path;
        return false;
      },
    });

    if (_path) {
      break;
    }
  }

  return _path;
}