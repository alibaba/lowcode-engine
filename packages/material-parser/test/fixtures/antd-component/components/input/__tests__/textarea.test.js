import React from 'react';
import { mount } from 'enzyme';
// eslint-disable-next-line import/no-unresolved
import Input from '..';
import focusTest from '../../../tests/shared/focusTest';
import calculateNodeHeight, { calculateNodeStyling } from '../calculateNodeHeight';
import { sleep } from '../../../tests/utils';

const { TextArea } = Input;

focusTest(TextArea);

describe('TextArea', () => {
  const originalGetComputedStyle = window.getComputedStyle;
  beforeAll(() => {
    Object.defineProperty(window, 'getComputedStyle', {
      value: node => ({
        getPropertyValue: prop => {
          if (prop === 'box-sizing') {
            return originalGetComputedStyle(node)[prop] || 'border-box';
          }
          return originalGetComputedStyle(node)[prop];
        },
      }),
    });
  });

  afterAll(() => {
    Object.defineProperty(window, 'getComputedStyle', {
      value: originalGetComputedStyle,
    });
  });

  it('should auto calculate height according to content length', async () => {
    const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

    const wrapper = mount(
      <TextArea value="" readOnly autoSize={{ minRows: 2, maxRows: 6 }} wrap="off" />,
    );
    const mockFunc = jest.spyOn(wrapper.instance().resizableTextArea, 'resizeTextarea');
    wrapper.setProps({ value: '1111\n2222\n3333' });
    await sleep(0);
    expect(mockFunc).toHaveBeenCalledTimes(1);
    wrapper.setProps({ value: '1111' });
    await sleep(0);
    expect(mockFunc).toHaveBeenCalledTimes(2);
    wrapper.update();
    expect(wrapper.find('textarea').props().style.overflow).toBeFalsy();

    expect(errorSpy).not.toHaveBeenCalled();
    errorSpy.mockRestore();
  });

  it('should support onPressEnter and onKeyDown', () => {
    const fakeHandleKeyDown = jest.fn();
    const fakeHandlePressEnter = jest.fn();
    const wrapper = mount(
      <TextArea onKeyDown={fakeHandleKeyDown} onPressEnter={fakeHandlePressEnter} />,
    );
    /** keyCode 65 is A */
    wrapper.find('textarea').simulate('keydown', { keyCode: 65 });
    expect(fakeHandleKeyDown).toHaveBeenCalledTimes(1);
    expect(fakeHandlePressEnter).toHaveBeenCalledTimes(0);

    /** keyCode 13 is Enter */
    wrapper.find('textarea').simulate('keydown', { keyCode: 13 });
    expect(fakeHandleKeyDown).toHaveBeenCalledTimes(2);
    expect(fakeHandlePressEnter).toHaveBeenCalledTimes(1);
  });

  it('should support disabled', () => {
    const wrapper = mount(<TextArea disabled />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('should support maxLength', () => {
    const wrapper = mount(<TextArea maxLength={10} />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('calculateNodeStyling works correctly', () => {
    const wrapper = document.createElement('textarea');
    wrapper.id = 'test';
    wrapper.wrap = 'wrap';
    calculateNodeStyling(wrapper, true);
    const value = calculateNodeStyling(wrapper, true);
    expect(value).toEqual({
      borderSize: 2,
      boxSizing: 'border-box',
      paddingSize: 4,
      sizingStyle:
        'letter-spacing:normal;line-height:normal;padding-top:2px;padding-bottom:2px;font-family:-webkit-small-control;font-weight:;font-size:;font-variant:;text-rendering:auto;text-transform:none;width:;text-indent:0;padding-left:2px;padding-right:2px;border-width:1px;box-sizing:border-box',
    });
  });

  it('boxSizing === "border-box"', () => {
    const wrapper = document.createElement('textarea');
    wrapper.style.boxSizing = 'border-box';
    const { height } = calculateNodeHeight(wrapper);
    expect(height).toBe(2);
  });

  it('boxSizing === "content-box"', () => {
    const wrapper = document.createElement('textarea');
    wrapper.style.boxSizing = 'content-box';
    const { height } = calculateNodeHeight(wrapper);
    expect(height).toBe(-4);
  });

  it('minRows or maxRows is not null', () => {
    const wrapper = document.createElement('textarea');
    expect(calculateNodeHeight(wrapper, 1, 1)).toEqual({
      height: 2,
      maxHeight: 9007199254740991,
      minHeight: 2,
      overflowY: undefined,
    });
    wrapper.style.boxSizing = 'content-box';
    expect(calculateNodeHeight(wrapper, 1, 1)).toEqual({
      height: -4,
      maxHeight: 9007199254740991,
      minHeight: -4,
      overflowY: undefined,
    });
  });

  it('when prop value not in this.props, resizeTextarea should be called', () => {
    const wrapper = mount(<TextArea aria-label="textarea" />);
    const resizeTextarea = jest.spyOn(wrapper.instance().resizableTextArea, 'resizeTextarea');
    wrapper.find('textarea').simulate('change', 'test');
    expect(resizeTextarea).toHaveBeenCalled();
  });

  it('handleKeyDown', () => {
    const onPressEnter = jest.fn();
    const onKeyDown = jest.fn();
    const wrapper = mount(
      <TextArea onPressEnter={onPressEnter} onKeyDown={onKeyDown} aria-label="textarea" />,
    );
    wrapper.instance().handleKeyDown({ keyCode: 13 });
    expect(onPressEnter).toHaveBeenCalled();
    expect(onKeyDown).toHaveBeenCalled();
  });

  it('should trigger onResize', async () => {
    const onResize = jest.fn();
    const wrapper = mount(<TextArea onResize={onResize} autoSize />);
    await sleep(100);
    wrapper
      .find('ResizeObserver')
      .instance()
      .onResize([
        {
          target: {
            getBoundingClientRect() {
              return {};
            },
          },
        },
      ]);

    expect(onResize).toHaveBeenCalledWith(
      expect.objectContaining({
        width: expect.any(Number),
        height: expect.any(Number),
      }),
    );
  });
});

describe('TextArea allowClear', () => {
  it('should change type when click', () => {
    const wrapper = mount(<TextArea allowClear />);
    wrapper.find('textarea').simulate('change', { target: { value: '111' } });
    expect(wrapper.find('textarea').getDOMNode().value).toEqual('111');
    expect(wrapper.render()).toMatchSnapshot();
    wrapper.find('.ant-input-textarea-clear-icon').at(0).simulate('click');
    expect(wrapper.render()).toMatchSnapshot();
    expect(wrapper.find('textarea').getDOMNode().value).toEqual('');
  });

  it('should not show icon if value is undefined, null or empty string', () => {
    const wrappers = [null, undefined, ''].map(val => mount(<TextArea allowClear value={val} />));
    wrappers.forEach(wrapper => {
      expect(wrapper.find('textarea').getDOMNode().value).toEqual('');
      expect(wrapper.find('.ant-input-textarea-clear-icon-hidden').exists()).toBeTruthy();
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  it('should not show icon if defaultValue is undefined, null or empty string', () => {
    const wrappers = [null, undefined, ''].map(val =>
      mount(<TextArea allowClear defaultValue={val} />),
    );
    wrappers.forEach(wrapper => {
      expect(wrapper.find('textarea').getDOMNode().value).toEqual('');
      expect(wrapper.find('.ant-textarea-clear-icon').exists()).toEqual(false);
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  it('should trigger event correctly', () => {
    let argumentEventObject;
    let argumentEventObjectValue;
    const onChange = e => {
      argumentEventObject = e;
      argumentEventObjectValue = e.target.value;
    };
    const wrapper = mount(<TextArea allowClear defaultValue="111" onChange={onChange} />);
    wrapper.find('.ant-input-textarea-clear-icon').at(0).simulate('click');
    expect(argumentEventObject.type).toBe('click');
    expect(argumentEventObjectValue).toBe('');
    expect(wrapper.find('textarea').at(0).getDOMNode().value).toBe('');
  });

  it('should trigger event correctly on controlled mode', () => {
    let argumentEventObject;
    let argumentEventObjectValue;
    const onChange = e => {
      argumentEventObject = e;
      argumentEventObjectValue = e.target.value;
    };
    const wrapper = mount(<TextArea allowClear value="111" onChange={onChange} />);
    wrapper.find('.ant-input-textarea-clear-icon').at(0).simulate('click');
    expect(argumentEventObject.type).toBe('click');
    expect(argumentEventObjectValue).toBe('');
    expect(wrapper.find('textarea').at(0).getDOMNode().value).toBe('111');
  });

  it('should focus textarea after clear', () => {
    const wrapper = mount(<TextArea allowClear defaultValue="111" />, { attachTo: document.body });
    wrapper.find('.ant-input-textarea-clear-icon').at(0).simulate('click');
    expect(document.activeElement).toBe(wrapper.find('textarea').at(0).getDOMNode());
    wrapper.unmount();
  });

  it('should not support allowClear when it is disabled', () => {
    const wrapper = mount(<TextArea allowClear defaultValue="111" disabled />);
    expect(wrapper.find('.ant-input-textarea-clear-icon-hidden').exists()).toBeTruthy();
  });

  it('not block input when `value` is undefined', () => {
    const wrapper = mount(<Input value={undefined} />);
    wrapper.find('input').simulate('change', { target: { value: 'Bamboo' } });
    expect(wrapper.find('input').props().value).toEqual('Bamboo');

    // Controlled
    wrapper.setProps({ value: 'Light' });
    wrapper.find('input').simulate('change', { target: { value: 'Bamboo' } });
    expect(wrapper.find('input').props().value).toEqual('Light');
  });

  describe('click focus', () => {
    it('click outside should also get focus', () => {
      const wrapper = mount(<Input suffix={<span className="test-suffix" />} />);
      const onFocus = jest.spyOn(wrapper.find('input').instance(), 'focus');
      wrapper.find('.test-suffix').simulate('mouseUp');
      expect(onFocus).toHaveBeenCalled();
    });

    it('not get focus if out of component', () => {
      const wrapper = mount(<Input suffix={<span className="test-suffix" />} />);
      const onFocus = jest.spyOn(wrapper.find('input').instance(), 'focus');
      const ele = document.createElement('span');
      document.body.appendChild(ele);
      wrapper.find('.test-suffix').simulate('mouseUp', {
        target: ele,
      });
      expect(onFocus).not.toHaveBeenCalled();
      document.body.removeChild(ele);
    });
  });

  it('scroll to bottom when autoSize', async () => {
    const wrapper = mount(<Input.TextArea autoSize />, { attachTo: document.body });
    wrapper.find('textarea').simulate('focus');
    wrapper.find('textarea').getDOMNode().focus();
    const setSelectionRangeFn = jest.spyOn(
      wrapper.find('textarea').getDOMNode(),
      'setSelectionRange',
    );
    wrapper.find('textarea').simulate('input', { target: { value: '\n1' } });
    await sleep(100);
    expect(setSelectionRangeFn).toHaveBeenCalled();
    wrapper.unmount();
  });
});
