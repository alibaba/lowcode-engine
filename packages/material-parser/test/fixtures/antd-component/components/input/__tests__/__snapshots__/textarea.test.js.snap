// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea allowClear should change type when click 1`] = `
<span
  class="ant-input-affix-wrapper ant-input-affix-wrapper-textarea-with-clear-btn"
>
  <textarea
    class="ant-input"
  >
    111
  </textarea>
  <span
    aria-label="close-circle"
    class="anticon anticon-close-circle ant-input-textarea-clear-icon"
    role="button"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class=""
      data-icon="close-circle"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
      />
    </svg>
  </span>
</span>
`;

exports[`TextArea allowClear should change type when click 2`] = `
<span
  class="ant-input-affix-wrapper ant-input-affix-wrapper-textarea-with-clear-btn"
>
  <textarea
    class="ant-input"
  />
  <span
    aria-label="close-circle"
    class="anticon anticon-close-circle ant-input-textarea-clear-icon ant-input-textarea-clear-icon-hidden"
    role="button"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class=""
      data-icon="close-circle"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
      />
    </svg>
  </span>
</span>
`;

exports[`TextArea allowClear should not show icon if defaultValue is undefined, null or empty string 1`] = `
<span
  class="ant-input-affix-wrapper ant-input-affix-wrapper-textarea-with-clear-btn"
>
  <textarea
    class="ant-input"
  />
  <span
    aria-label="close-circle"
    class="anticon anticon-close-circle ant-input-textarea-clear-icon ant-input-textarea-clear-icon-hidden"
    role="button"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class=""
      data-icon="close-circle"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
      />
    </svg>
  </span>
</span>
`;

exports[`TextArea allowClear should not show icon if defaultValue is undefined, null or empty string 2`] = `
<span
  class="ant-input-affix-wrapper ant-input-affix-wrapper-textarea-with-clear-btn"
>
  <textarea
    class="ant-input"
  />
  <span
    aria-label="close-circle"
    class="anticon anticon-close-circle ant-input-textarea-clear-icon ant-input-textarea-clear-icon-hidden"
    role="button"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class=""
      data-icon="close-circle"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
      />
    </svg>
  </span>
</span>
`;

exports[`TextArea allowClear should not show icon if defaultValue is undefined, null or empty string 3`] = `
<span
  class="ant-input-affix-wrapper ant-input-affix-wrapper-textarea-with-clear-btn"
>
  <textarea
    class="ant-input"
  />
  <span
    aria-label="close-circle"
    class="anticon anticon-close-circle ant-input-textarea-clear-icon ant-input-textarea-clear-icon-hidden"
    role="button"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class=""
      data-icon="close-circle"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
      />
    </svg>
  </span>
</span>
`;

exports[`TextArea allowClear should not show icon if value is undefined, null or empty string 1`] = `
<span
  class="ant-input-affix-wrapper ant-input-affix-wrapper-textarea-with-clear-btn"
>
  <textarea
    class="ant-input"
  />
  <span
    aria-label="close-circle"
    class="anticon anticon-close-circle ant-input-textarea-clear-icon ant-input-textarea-clear-icon-hidden"
    role="button"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class=""
      data-icon="close-circle"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
      />
    </svg>
  </span>
</span>
`;

exports[`TextArea allowClear should not show icon if value is undefined, null or empty string 2`] = `
<span
  class="ant-input-affix-wrapper ant-input-affix-wrapper-textarea-with-clear-btn"
>
  <textarea
    class="ant-input"
  />
  <span
    aria-label="close-circle"
    class="anticon anticon-close-circle ant-input-textarea-clear-icon ant-input-textarea-clear-icon-hidden"
    role="button"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class=""
      data-icon="close-circle"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
      />
    </svg>
  </span>
</span>
`;

exports[`TextArea allowClear should not show icon if value is undefined, null or empty string 3`] = `
<span
  class="ant-input-affix-wrapper ant-input-affix-wrapper-textarea-with-clear-btn"
>
  <textarea
    class="ant-input"
  />
  <span
    aria-label="close-circle"
    class="anticon anticon-close-circle ant-input-textarea-clear-icon ant-input-textarea-clear-icon-hidden"
    role="button"
    tabindex="-1"
  >
    <svg
      aria-hidden="true"
      class=""
      data-icon="close-circle"
      fill="currentColor"
      focusable="false"
      height="1em"
      viewBox="64 64 896 896"
      width="1em"
    >
      <path
        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
      />
    </svg>
  </span>
</span>
`;

exports[`TextArea should support disabled 1`] = `
<textarea
  class="ant-input ant-input-disabled"
  disabled=""
/>
`;

exports[`TextArea should support maxLength 1`] = `
<textarea
  class="ant-input"
  maxlength="10"
/>
`;
